File: packages/sdk-go/semantic_cache_test.go
package rateguard

import (
	"context"
	"encoding/json"
	"errors"
	"io"
	"net/http"
	"strings"
	"sync"
	"testing"
	"time"
)

// stubEmbedder maps known phrases to fixed vectors so similarity is
// deterministic in tests. Unknown text embeds to the zero vector (never
// matches anything, cosine similarity of a zero vector is defined as 0).
type stubEmbedder struct {
	mu    sync.Mutex
	calls int
	// vectors keyed by substring match against the prompt text.
	vectors map[string][]float32
	err     error
}

func (e *stubEmbedder) Embed(_ context.Context, text string) ([]float32, error) {
	e.mu.Lock()
	e.calls++
	e.mu.Unlock()
	if e.err != nil {
		return nil, e.err
	}
	for substr, vec := range e.vectors {
		if strings.Contains(text, substr) {
			return vec, nil
		}
	}
	return []float32{0, 0, 0}, nil
}

func (e *stubEmbedder) callCount() int {
	e.mu.Lock()
	defer e.mu.Unlock()
	return e.calls
}

func TestCosineSimilarity(t *testing.T) {
	cases := []struct {
		name string
		a, b []float32
		want float64
	}{
		{"identical", []float32{1, 0, 0}, []float32{1, 0, 0}, 1.0},
		{"orthogonal", []float32{1, 0}, []float32{0, 1}, 0.0},
		{"opposite", []float32{1, 0}, []float32{-1, 0}, -1.0},
		{"mismatched length", []float32{1, 0}, []float32{1, 0, 0}, 0.0},
		{"empty", nil, nil, 0.0},
		{"zero vector", []float32{0, 0}, []float32{1, 1}, 0.0},
	}
	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			got := cosineSimilarity(tc.a, tc.b)
			if diff := got - tc.want; diff > 1e-9 || diff < -1e-9 {
				t.Fatalf("cosineSimilarity(%v, %v) = %v, want %v", tc.a, tc.b, got, tc.want)
			}
		})
	}
}

func TestSemanticCacheHitAboveThreshold(t *testing.T) {
	clock := &fakeLimiterClock{now: time.Date(2026, 7, 5, 12, 0, 0, 0, time.UTC)}
	c := newSemanticCache(SemanticCacheOptions{SimilarityThreshold: 0.9}, clock)

	stored := cachedLLMResponse{statusCode: 200, body: []byte(`{"ok":true}`)}
	c.store("openai:gpt-4o", []float32{1, 0, 0}, stored)

	got, ok := c.lookup("openai:gpt-4o", []float32{1, 0.01, 0})
	if !ok {
		t.Fatal("expected a hit for a near-identical embedding")
	}
	if string(got.body) != string(stored.body) {
		t.Fatalf("got body %q, want %q", got.body, stored.body)
	}
}

func TestSemanticCacheMissBelowThreshold(t *testing.T) {
	clock := &fakeLimiterClock{now: time.Date(2026, 7, 5, 12, 0, 0, 0, time.UTC)}
	c := newSemanticCache(SemanticCacheOptions{SimilarityThreshold: 0.95}, clock)

	c.store("openai:gpt-4o", []float32{1, 0, 0}, cachedLLMResponse{statusCode: 200})

	if _, ok := c.lookup("openai:gpt-4o", []float32{0, 1, 0}); ok {
		t.Fatal("orthogonal embedding must not hit")
	}
}

func TestSemanticCacheScopeIsolation(t *testing.T) {
	clock := &fakeLimiterClock{now: time.Date(2026, 7, 5, 12, 0, 0, 0, time.UTC)}
	c := newSemanticCache(SemanticCacheOptions{SimilarityThreshold: 0.9}, clock)

	c.store("openai:gpt-4o", []float32{1, 0, 0}, cachedLLMResponse{statusCode: 200, body: []byte("gpt-4o")})

	if _, ok := c.lookup("anthropic:claude-opus-4-5", []float32{1, 0, 0}); ok {
		t.Fatal("identical embedding in a different provider:model scope must not hit")
	}
}

func TestSemanticCacheExpiry(t *testing.T) {
	clock := &fakeLimiterClock{now: time.Date(2026, 7, 5, 12, 0, 0, 0, time.UTC)}
	c := newSemanticCache(SemanticCacheOptions{SimilarityThreshold: 0.9, TTL: time.Minute}, clock)

	c.store("openai:gpt-4o", []float32{1, 0, 0}, cachedLLMResponse{statusCode: 200})

	if _, ok := c.lookup("openai:gpt-4o", []float32{1, 0, 0}); !ok {
		t.Fatal("expected a hit before expiry")
	}

	clock.advance(2 * time.Minute)
	if _, ok := c.lookup("openai:gpt-4o", []float32{1, 0, 0}); ok {
		t.Fatal("expired entry must not hit")
	}
}

func TestSemanticCacheBoundedEviction(t *testing.T) {
	clock := &fakeLimiterClock{now: time.Date(2026, 7, 5, 12, 0, 0, 0, time.UTC)}
	c := newSemanticCache(SemanticCacheOptions{SimilarityThreshold: 0.999, MaxEntriesPerScope: 3}, clock)

	// Four orthogonal-ish entries in a 4-D space, each store distinct.
	c.store("s", []float32{1, 0, 0, 0}, cachedLLMResponse{body: []byte("a")})
	c.store("s", []float32{0, 1, 0, 0}, cachedLLMResponse{body: []byte("b")})
	c.store("s", []float32{0, 0, 1, 0}, cachedLLMResponse{body: []byte("c")})
	c.store("s", []float32{0, 0, 0, 1}, cachedLLMResponse{body: []byte("d")})

	// Oldest ("a") must have been evicted; newest ("d") must still be present.
	if _, ok := c.lookup("s", []float32{1, 0, 0, 0}); ok {
		t.Fatal("oldest entry should have been evicted at capacity 3")
	}
	if _, ok := c.lookup("s", []float32{0, 0, 0, 1}); !ok {
		t.Fatal("newest entry must survive eviction")
	}
}

func TestPromptTextFromRequestBodyOpenAIShape(t *testing.T) {
	body := []byte(`{"model":"gpt-4o","messages":[
		{"role":"system","content":"You are terse."},
		{"role":"user","content":"What is 2+2?"}
	]}`)
	got := promptTextFromRequestBody(body)
	if !strings.Contains(got, "You are terse.") || !strings.Contains(got, "What is 2+2?") {
		t.Fatalf("prompt extraction missed content: %q", got)
	}
}

func TestPromptTextFromRequestBodyMultimodalParts(t *testing.T) {
	body := []byte(`{"messages":[{"role":"user","content":[
		{"type":"text","text":"describe this"},
		{"type":"image_url","image_url":{"url":"https://example.com/x.png"}}
	]}]}`)
	got := promptTextFromRequestBody(body)
	if !strings.Contains(got, "describe this") {
		t.Fatalf("expected text part extracted, got %q", got)
	}
	if strings.Contains(got, "example.com") {
		t.Fatalf("image part must be ignored, got %q", got)
	}
}

func TestPromptTextFromRequestBodyAnthropicSystemField(t *testing.T) {
	body := []byte(`{"model":"claude-opus-4-5","system":"Be concise.","messages":[
		{"role":"user","content":"hello"}
	]}`)
	got := promptTextFromRequestBody(body)
	if !strings.Contains(got, "Be concise.") || !strings.Contains(got, "hello") {
		t.Fatalf("expected system + message content, got %q", got)
	}
}

func TestIsStreamingRequestBody(t *testing.T) {
	if isStreamingRequestBody([]byte(`{"model":"gpt-4o","stream":true}`)) != true {
		t.Fatal("stream:true must be detected")
	}
	if isStreamingRequestBody([]byte(`{"model":"gpt-4o","stream":false}`)) != false {
		t.Fatal("stream:false must not be detected as streaming")
	}
	if isStreamingRequestBody([]byte(`{"model":"gpt-4o"}`)) != false {
		t.Fatal("absent stream field must default to non-streaming")
	}
	if isStreamingRequestBody(nil) != false {
		t.Fatal("empty body must not be treated as streaming")
	}
}

// ── Transport-level integration ──

type fixedTransport struct {
	mu    sync.Mutex
	calls int
	fn    func(*http.Request) (*http.Response, error)
}

func (f *fixedTransport) RoundTrip(req *http.Request) (*http.Response, error) {
	f.mu.Lock()
	f.calls++
	f.mu.Unlock()
	return f.fn(req)
}

func (f *fixedTransport) callCount() int {
	f.mu.Lock()
	defer f.mu.Unlock()
	return f.calls
}

func jsonOKResponse(req *http.Request, payload string) *http.Response {
	header := http.Header{}
	header.Set("Content-Type", "application/json")
	return &http.Response{
		StatusCode:    http.StatusOK,
		Header:        header,
		Body:          io.NopCloser(strings.NewReader(payload)),
		ContentLength: int64(len(payload)),
		Request:       req,
	}
}

func openAIChatRequest(t *testing.T, prompt string) *http.Request {
	t.Helper()
	body, err := json.Marshal(map[string]any{
		"model":    "gpt-4o",
		"messages": []map[string]string{{"role": "user", "content": prompt}},
	})
	if err != nil {
		t.Fatal(err)
	}
	req, err := http.NewRequest(http.MethodPost, "https://api.openai.com/v1/chat/completions", strings.NewReader(string(body)))
	if err != nil {
		t.Fatal(err)
	}
	return req
}

func TestOutboundSemanticCacheHitSkipsNetwork(t *testing.T) {
	upstream := &fixedTransport{fn: func(req *http.Request) (*http.Response, error) {
		return jsonOKResponse(req, `{"choices":[{"message":{"content":"hi"}}],"usage":{"prompt_tokens":5,"completion_tokens":3,"total_tokens":8}}`), nil
	}}
	embedder := &stubEmbedder{vectors: map[string][]float32{"capital of france": {1, 0, 0}}}

	sdk := New(Config{Preset: "standard"})
	client := sdk.WrapClient(&http.Client{Transport: upstream}, OutboundOptions{
		SemanticCache: &SemanticCacheOptions{Embedder: embedder, SimilarityThreshold: 0.9},
	})

	req1 := openAIChatRequest(t, "what is the capital of france?")
	resp1, err := client.Do(req1)
	if err != nil {
		t.Fatal(err)
	}
	resp1.Body.Close()
	if resp1.Header.Get("X-RateGuard-Cache") == "hit" {
		t.Fatal("first call must be a real miss")
	}
	if upstream.callCount() != 1 {
		t.Fatalf("expected 1 upstream call after first request, got %d", upstream.callCount())
	}

	req2 := openAIChatRequest(t, "tell me the capital of france")
	resp2, err := client.Do(req2)
	if err != nil {
		t.Fatal(err)
	}
	body2, _ := io.ReadAll(resp2.Body)
	resp2.Body.Close()

	if resp2.Header.Get("X-RateGuard-Cache") != "hit" {
		t.Fatal("similar second prompt must be served from cache")
	}
	if upstream.callCount() != 1 {
		t.Fatalf("cache hit must not reach upstream: calls = %d", upstream.callCount())
	}
	if !strings.Contains(string(body2), `"content":"hi"`) {
		t.Fatalf("cached body mismatch: %s", body2)
	}
}

func TestOutboundSemanticCacheStreamingBypassesCache(t *testing.T) {
	calls := 0
	upstream := &fixedTransport{fn: func(req *http.Request) (*http.Response, error) {
		calls++
		return jsonOKResponse(req, `data: [DONE]`), nil
	}}
	embedder := &stubEmbedder{}

	sdk := New(Config{Preset: "standard"})
	client := sdk.WrapClient(&http.Client{Transport: upstream}, OutboundOptions{
		SemanticCache: &SemanticCacheOptions{Embedder: embedder},
	})

	body, _ := json.Marshal(map[string]any{
		"model": "gpt-4o", "stream": true,
		"messages": []map[string]string{{"role": "user", "content": "hello"}},
	})
	req, _ := http.NewRequest(http.MethodPost, "https://api.openai.com/v1/chat/completions", strings.NewReader(string(body)))
	req2, _ := http.NewRequest(http.MethodPost, "https://api.openai.com/v1/chat/completions", strings.NewReader(string(body)))

	if _, err := client.Do(req); err != nil {
		t.Fatal(err)
	}
	if _, err := client.Do(req2); err != nil {
		t.Fatal(err)
	}

	if embedder.callCount() != 0 {
		t.Fatalf("streaming requests must never be embedded, got %d calls", embedder.callCount())
	}
	if calls != 2 {
		t.Fatalf("streaming requests must always hit the network, got %d calls", calls)
	}
}

func TestOutboundSemanticCacheEmbedderFailureDegradesToRealCall(t *testing.T) {
	upstream := &fixedTransport{fn: func(req *http.Request) (*http.Response, error) {
		return jsonOKResponse(req, `{"ok":true}`), nil
	}}
	embedder := &stubEmbedder{err: errors.New("embedding service down")}

	sdk := New(Config{Preset: "standard"})
	client := sdk.WrapClient(&http.Client{Transport: upstream}, OutboundOptions{
		SemanticCache: &SemanticCacheOptions{Embedder: embedder},
	})

	resp, err := client.Do(openAIChatRequest(t, "hello"))
	if err != nil {
		t.Fatalf("embedder failure must not fail the request: %v", err)
	}
	resp.Body.Close()
	if upstream.callCount() != 1 {
		t.Fatalf("expected the real call to proceed, got %d upstream calls", upstream.callCount())
	}
}

func TestOutboundSemanticCacheNeverCachesNonOKResponses(t *testing.T) {
	upstream := &fixedTransport{fn: func(req *http.Request) (*http.Response, error) {
		header := http.Header{}
		header.Set("Content-Type", "application/json")
		payload := `{"error":"upstream unavailable"}`
		return &http.Response{
			StatusCode:    http.StatusInternalServerError,
			Header:        header,
			Body:          io.NopCloser(strings.NewReader(payload)),
			ContentLength: int64(len(payload)),
			Request:       req,
		}, nil
	}}
	embedder := &stubEmbedder{vectors: map[string][]float32{"hello": {1, 0, 0}}}

	sdk := New(Config{Preset: "standard"})
	client := sdk.WrapClient(&http.Client{Transport: upstream}, OutboundOptions{
		SemanticCache: &SemanticCacheOptions{Embedder: embedder, SimilarityThreshold: 0.9},
		Chain:         nil, // no fallback — the 500 must surface, not be retried away
	})

	for i := 0; i < 2; i++ {
		resp, err := client.Do(openAIChatRequest(t, "hello there"))
		if err != nil {
			t.Fatal(err)
		}
		resp.Body.Close()
	}

	if upstream.callCount() != 2 {
		t.Fatalf("a 500 response must never be cached — expected 2 upstream calls, got %d", upstream.callCount())
	}
}

File: packages/sdk-node/test/semantic-cache.test.ts
import { describe, expect, it } from 'vitest';
import { RateGuard } from '../src/index.js';
import {
  SemanticCache,
  cosineSimilarity,
  isStreamingRequestBody,
  promptTextFromRequestBody,
  type Embedder,
} from '../src/core/semantic-cache.js';

function fakeClock(startMs: number) {
  let now = startMs;
  return {
    clock: { now: () => now },
    advance: (ms: number) => {
      now += ms;
    },
  };
}

const openAIBody = (model: string, prompt: number, completion: number): string =>
  JSON.stringify({
    id: 'cmpl-1',
    model,
    choices: [{ message: { content: 'hi' } }],
    usage: { prompt_tokens: prompt, completion_tokens: completion, total_tokens: prompt + completion },
  });

describe('cosineSimilarity', () => {
  it('is 1 for identical vectors, 0 for orthogonal, and 0 for mismatched/empty/zero-norm inputs', () => {
    expect(cosineSimilarity([1, 0], [1, 0])).toBeCloseTo(1, 10);
    expect(cosineSimilarity([1, 0], [0, 1])).toBeCloseTo(0, 10);
    expect(cosineSimilarity([], [])).toBe(0);
    expect(cosineSimilarity([1, 2], [1, 2, 3])).toBe(0);
    expect(cosineSimilarity([0, 0], [1, 1])).toBe(0);
  });
});

describe('isStreamingRequestBody', () => {
  it('detects stream:true and treats everything else (including malformed JSON) as non-streaming', () => {
    expect(isStreamingRequestBody('{"stream":true}')).toBe(true);
    expect(isStreamingRequestBody('{"stream":false}')).toBe(false);
    expect(isStreamingRequestBody('{}')).toBe(false);
    expect(isStreamingRequestBody('')).toBe(false);
    expect(isStreamingRequestBody('not json')).toBe(false);
  });
});

describe('promptTextFromRequestBody', () => {
  it('extracts system + message text and ignores non-text content parts', () => {
    const body = JSON.stringify({
      system: 'You are helpful.',
      messages: [
        { role: 'user', content: 'What is the capital of France?' },
        {
          role: 'assistant',
          content: [
            { type: 'text', text: 'Paris.' },
            { type: 'image', url: 'https://example.com/x.png' },
          ],
        },
      ],
    });
    expect(promptTextFromRequestBody(body)).toBe(
      'system: You are helpful.\nuser: What is the capital of France?\nassistant: Paris.\n',
    );
  });

  it('returns empty string for non-JSON, empty, or content-less bodies', () => {
    expect(promptTextFromRequestBody('')).toBe('');
    expect(promptTextFromRequestBody('not json')).toBe('');
    expect(promptTextFromRequestBody('{}')).toBe('');
  });
});

describe('SemanticCache', () => {
  it('hits on a similar-enough embedding within threshold and misses below it', () => {
    const embedder: Embedder = { embed: async () => [] }; // unused: lookup/store take explicit vectors
    const cache = new SemanticCache({ embedder, similarityThreshold: 0.9 }, { now: () => 0 });

    cache.store('openai:gpt-4o', [1, 0, 0], { status: 200, headers: {}, body: 'cached-response' });

    // cosine([1,0,0], [0.99, 0.14, 0]) ~= 0.99 -> above threshold -> hit
    expect(cache.lookup('openai:gpt-4o', [0.99, 0.14, 0])?.body).toBe('cached-response');

    // orthogonal vector -> miss
    expect(cache.lookup('openai:gpt-4o', [0, 1, 0])).toBeUndefined();

    // different scope entirely -> miss regardless of similarity
    expect(cache.lookup('anthropic:claude-sonnet-4', [1, 0, 0])).toBeUndefined();
  });

  it('expires entries lazily and never serves an expired entry', () => {
    const { clock, advance } = fakeClock(0);
    const embedder: Embedder = { embed: async () => [] };
    const cache = new SemanticCache({ embedder, ttlMs: 1_000 }, clock);

    cache.store('scope', [1, 0], { status: 200, headers: {}, body: 'fresh' });
    expect(cache.lookup('scope', [1, 0])?.body).toBe('fresh');

    advance(1_001);
    expect(cache.lookup('scope', [1, 0])).toBeUndefined();
  });

  it('evicts oldest entries first once maxEntriesPerScope is exceeded', () => {
    const embedder: Embedder = { embed: async () => [] };
    const cache = new SemanticCache({ embedder, maxEntriesPerScope: 2, similarityThreshold: 0.99 }, { now: () => 0 });

    cache.store('scope', [1, 0, 0], { status: 200, headers: {}, body: 'first' });
    cache.store('scope', [0, 1, 0], { status: 200, headers: {}, body: 'second' });
    cache.store('scope', [0, 0, 1], { status: 200, headers: {}, body: 'third' }); // evicts "first"

    expect(cache.lookup('scope', [1, 0, 0])).toBeUndefined();
    expect(cache.lookup('scope', [0, 1, 0])?.body).toBe('second');
    expect(cache.lookup('scope', [0, 0, 1])?.body).toBe('third');
  });
});

/** Deterministic fake embedder: 26-dim letter-frequency vector (case-insensitive). */
class LetterFrequencyEmbedder implements Embedder {
  async embed(text: string): Promise<number[]> {
    const vec = new Array(26).fill(0);
    for (const ch of text.toLowerCase()) {
      const code = ch.charCodeAt(0) - 97;
      if (code >= 0 && code < 26) {
        vec[code] += 1;
      }
    }
    return vec;
  }
}

describe('wrapFetch semanticCache wiring', () => {
  it('serves a cache hit for a similar-enough prompt without calling the network again', async () => {
    let networkCalls = 0;
    const rg = new RateGuard({ preset: 'dev' });
    const wrapped = rg.wrapFetch({
      semanticCache: { embedder: new LetterFrequencyEmbedder(), similarityThreshold: 0.95 },
      fetch: (async () => {
        networkCalls++;
        return new Response(openAIBody('gpt-4o', 10, 5), { status: 200, headers: { 'content-type': 'application/json' } });
      }) as typeof fetch,
    });

    const first = await wrapped('https://api.openai.com/v1/chat/completions', {
      method: 'POST',
      body: JSON.stringify({ model: 'gpt-4o', messages: [{ role: 'user', content: 'Tell me a fun fact about octopuses' }] }),
    });
    expect(first.status).toBe(200);
    expect(networkCalls).toBe(1);
    const firstText = await first.text();

    // Same letters (only case/punctuation differ, which the fake embedder
    // ignores/normalizes) -> identical embedding under this fixture -> hit.
    const second = await wrapped('https://api.openai.com/v1/chat/completions', {
      method: 'POST',
      body: JSON.stringify({ model: 'gpt-4o', messages: [{ role: 'user', content: 'tell me a fun fact about octopuses!' }] }),
    });
    expect(second.status).toBe(200);
    expect(networkCalls).toBe(1); // no additional network call
    expect(second.headers.get('x-rateguard-cache')).toBe('hit');
    expect(await second.text()).toBe(firstText);
  });

  it('misses the cache for a dissimilar prompt and stores it as a new entry', async () => {
    let networkCalls = 0;
    const rg = new RateGuard({ preset: 'dev' });
    const wrapped = rg.wrapFetch({
      semanticCache: { embedder: new LetterFrequencyEmbedder(), similarityThreshold: 0.95 },
      fetch: (async () => {
        networkCalls++;
        return new Response(openAIBody('gpt-4o', 10, 5), { status: 200, headers: { 'content-type': 'application/json' } });
      }) as typeof fetch,
    });

    await wrapped('https://api.openai.com/v1/chat/completions', {
      method: 'POST',
      body: JSON.stringify({ model: 'gpt-4o', messages: [{ role: 'user', content: 'Tell me a fun fact about octopuses' }] }),
    });
    expect(networkCalls).toBe(1);

    const unrelated = await wrapped('https://api.openai.com/v1/chat/completions', {
      method: 'POST',
      body: JSON.stringify({ model: 'gpt-4o', messages: [{ role: 'user', content: 'Summarize quantum entanglement in Bell test experiments' }] }),
    });
    expect(unrelated.status).toBe(200);
    expect(networkCalls).toBe(2); // dissimilar prompt -> real call, not a cache hit
    expect(unrelated.headers.get('x-rateguard-cache')).toBeNull();
  });

  // Reproduces a real gap: the scope key was computed from the REQUESTED
  // provider/model before the call ran, but a fallback (retarget) never
  // mutates the caller's `call` object — it builds an entirely new one for
  // the recursive attempt. Caching a fallback answer under the pre-fallback
  // scope meant a later request that reached the ORIGINAL (now-recovered)
  // provider could get served the FALLBACK provider's stale, mislabeled
  // answer instead of a fresh real one.
  it('scopes a fallback response to the provider that actually served it, not the one requested', async () => {
    let primaryShouldFail = true;
    let primaryCalls = 0;
    let fallbackCalls = 0;
    const mockFetch = (async (input: RequestInfo | URL, init?: RequestInit) => {
      const body = typeof init?.body === 'string' ? init.body : '';
      if (body.includes('deepseek-chat')) {
        fallbackCalls++;
        return new Response(openAIBody('deepseek-chat', 10, 5), { status: 200, headers: { 'content-type': 'application/json' } });
      }
      primaryCalls++;
      if (primaryShouldFail) {
        return new Response('{"error":{"message":"rate limited"}}', { status: 429 });
      }
      return new Response(openAIBody('gpt-4o', 10, 5), { status: 200, headers: { 'content-type': 'application/json' } });
    }) as typeof fetch;

    const rg = new RateGuard({ preset: 'dev' });
    const wrapped = rg.wrapFetch({
      fetch: mockFetch,
      chain: [{ name: 'deepseek', model: 'deepseek-chat', baseURL: 'https://api.deepseek.com/v1', weight: 0 }],
      semanticCache: { embedder: new LetterFrequencyEmbedder(), similarityThreshold: 0.95 },
    });

    // Request 1: primary is down, falls back to deepseek. Gets cached —
    // the fix requires it be cached under deepseek's scope, not openai's.
    const prompt = 'what is the capital of france';
    const first = await wrapped('https://api.openai.com/v1/chat/completions', {
      method: 'POST',
      body: JSON.stringify({ model: 'gpt-4o', messages: [{ role: 'user', content: prompt }] }),
    });
    expect(first.headers.get('x-rateguard-fallback')).toBe('true');
    expect(primaryCalls).toBe(1);
    expect(fallbackCalls).toBe(1);

    // Primary has recovered. Request 2 (IDENTICAL prompt -> identical
    // embedding, guaranteeing a cache match if scope agrees) must NOT be
    // served deepseek's cached answer as if it were openai's — it must
    // reach the network and get a fresh, real openai response. If the bug
    // were present (cached under "openai:gpt-4o"), this would be a
    // wrongful cache hit: primaryCalls would stay at 1 and the body would
    // still say "deepseek-chat".
    primaryShouldFail = false;
    const second = await wrapped('https://api.openai.com/v1/chat/completions', {
      method: 'POST',
      body: JSON.stringify({ model: 'gpt-4o', messages: [{ role: 'user', content: prompt }] }),
    });

    expect(second.headers.get('x-rateguard-cache')).not.toBe('hit');
    expect(primaryCalls).toBe(2);
    const secondText = await second.text();
    expect(secondText).toContain('"model":"gpt-4o"');
  });

  it('never caches or serves streaming requests', async () => {
    let networkCalls = 0;
    // Generous token budget: the SSE usage scan that commits/releases the
    // outbound reservation runs asynchronously in the background (a
    // fire-and-forget tee), so two back-to-back streaming calls can have
    // overlapping in-flight reservations. A small budget would make the
    // second call's reservation legitimately contend with the first's —
    // a real budget-serialization behavior, but not what this test is
    // about (streaming bypasses the cache), so give it enough headroom.
    const rg = new RateGuard({ preset: 'dev', tokenBudget: { hourLimit: 1_000_000 } });
    const wrapped = rg.wrapFetch({
      semanticCache: { embedder: new LetterFrequencyEmbedder(), similarityThreshold: 0.95 },
      fetch: (async () => {
        networkCalls++;
        return new Response('data: [DONE]\n', { status: 200, headers: { 'content-type': 'text/event-stream' } });
      }) as typeof fetch,
    });

    const body = JSON.stringify({
      model: 'gpt-4o',
      stream: true,
      messages: [{ role: 'user', content: 'Tell me a fun fact about octopuses' }],
    });

    await wrapped('https://api.openai.com/v1/chat/completions', { method: 'POST', body });
    await wrapped('https://api.openai.com/v1/chat/completions', { method: 'POST', body });

    expect(networkCalls).toBe(2); // every streaming call hits the network — the cache is never consulted
  });

  it('does not embed or cache when no semanticCache option is configured', async () => {
    let networkCalls = 0;
    const rg = new RateGuard({ preset: 'dev' });
    const wrapped = rg.wrapFetch({
      fetch: (async () => {
        networkCalls++;
        return new Response(openAIBody('gpt-4o', 10, 5), { status: 200, headers: { 'content-type': 'application/json' } });
      }) as typeof fetch,
    });

    const body = JSON.stringify({ model: 'gpt-4o', messages: [{ role: 'user', content: 'same prompt' }] });
    await wrapped('https://api.openai.com/v1/chat/completions', { method: 'POST', body });
    await wrapped('https://api.openai.com/v1/chat/completions', { method: 'POST', body });

    expect(networkCalls).toBe(2); // no cache configured -> always a real call
  });
});
